app/src/main/java/com/lagradost/cloudstream3/syncproviders/providers/GoogleDriveApi.kt

```diff
@@ -0,0 +1,418 @@
+import androidx.browser.customtabs.CustomTabsIntent
+import androidx.fragment.app.FragmentActivity
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.google.api.client.auth.oauth2.AuthorizationCodeFlow
+import com.google.api.client.auth.oauth2.Credential
+import com.google.api.client.auth.oauth2.TokenResponse
+import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.http.FileContent
+import com.google.api.client.json.gson.GsonFactory
+import com.google.api.client.util.store.MemoryDataStoreFactory
+import com.google.api.services.drive.Drive
+import com.google.api.services.drive.DriveScopes
+import com.google.api.services.drive.model.File
+import com.lagradost.cloudstream3.AcraApplication
+import com.lagradost.cloudstream3.CommonActivity
+import com.lagradost.cloudstream3.R
+import com.lagradost.cloudstream3.syncproviders.AuthAPI
+import com.lagradost.cloudstream3.syncproviders.BackupAPI
+import com.lagradost.cloudstream3.syncproviders.InAppOAuth2API
+import com.lagradost.cloudstream3.syncproviders.InAppOAuth2APIManager
+import com.lagradost.cloudstream3.utils.AppUtils.toJson
+import com.lagradost.cloudstream3.utils.BackupUtils
+import com.lagradost.cloudstream3.utils.BackupUtils.getBackup
+import com.lagradost.cloudstream3.utils.BackupUtils.restore
+import com.lagradost.cloudstream3.utils.Coroutines.ioSafe
+import com.lagradost.cloudstream3.utils.DataStore
+import com.lagradost.cloudstream3.utils.DataStore.removeKey
+import kotlinx.coroutines.Job
+import java.io.InputStream
+import java.util.*
+
+
+// improvements and ideas
+//  - add option to use proper oauth through google services one tap - would need google console project on behalf of cloudstream
+//  - encrypt data on drive - meh
+//  - choose what should be synced
+//  - having a button to run sync would be nice - its really just for user to feel "safe"
+//  - having two or more devices in use at same time results in racing conditions
+//  - restoring backup should update view models - maybe just first fragment is enough
+//  - move "https://chiff.github.io/cloudstream-sync/google-drive"  to cs3 repo
+class GoogleDriveApi(index: Int) :
+    InAppOAuth2APIManager(index),
+    BackupAPI<InAppOAuth2API.LoginData> {
+    /////////////////////////////////////////
+    /////////////////////////////////////////
+    // Setup
+    override val key = "gdrive"
+    override val redirectUrl = "oauth/google-drive"
+
+    override val idPrefix = "gdrive"
+    override val name = "Google Drive"
+    override val icon = R.drawable.ic_baseline_add_to_drive_24
+
+    override val requiresFilename = true
+    override val requiresSecret = true
+    override val requiresClientId = true
+    override val defaultFilenameValue = "cloudstreamapp-sync-file"
+    override val defaultRedirectUrl =   "https://chiff.github.io/cloudstream-sync/google-drive"
```

Comment:
I assume Google doesn't allow to directly set the redirect url to `cloudstreamapp://oauth/google-drive`?

Reply:
Tldr; Yes you are right. 

Took a while to figure this out. There are two viable approaches I am aware of. 

Easiest for the user would be for cloudstream to create google console project, create an android oauth client and we would use in app loggin which is really smooth to use a looks really cool. However this may need to pay some kind of subscription because there is a quota for free. Also, there may be something like a privacy concern for users.

The other way (the one i implemented) is to create web client which allows redirects only to https origins and that's the reason for this redirect url. I tried deep links for smthn like https://my-non-existent.url/cloudstream/etc but I wasn't able to make it work. I think it is good compromise to create redirect url with cloudstream's gh-pages where source code is still visible for those concerned. Again, this method may seem more plausible for privacy concerned users as they have their own project under control.

Note: I will also create a guide to create projects and some explaining what it does. Im busy for next week and come back to implementation later the other on. I will still be able to answer any questions and if there is a need for 
a conversation I created feature request in discord called "bidirectional device sync"